File: dotnet/DemoApp/Core.Utilities/Filters/FunctionInvocationLoggingFilter.cs
using Microsoft.SemanticKernel;
using Microsoft.Extensions.Logging;
using System.Text.Json;

namespace Core.Utilities.Filters;

public sealed class FunctionInvocationLoggingFilter(ILogger logger)
    : IFunctionInvocationFilter
{
    public async Task OnFunctionInvocationAsync(
        FunctionInvocationContext context,
        Func<FunctionInvocationContext, Task> next)
    {
        if (logger.IsEnabled(LogLevel.Trace))
        {
            logger.LogTrace("function invoking: {Name}", context.Function.Name);
            logger.LogTrace("arguments: {Arguments}", JsonSerializer.Serialize(context.Arguments));
        }

        await next(context);

        if (logger.IsEnabled(LogLevel.Trace))
        {
            logger.LogTrace("function invoked: {Name}", context.Function.Name);
            logger.LogTrace("result: {Result}", context.Result);
        }
    }
}

File: python/DemoApp/core/utilities/filters/functioninvocationloggingfilter.py
import json
import logging
from semantic_kernel.filters.functions.function_invocation_context import FunctionInvocationContext
from collections.abc import Callable, Coroutine
from typing import Any


class FunctionInvocationLoggingFilter:
    def __init__(self, logger: logging.Logger) -> None:
        self.logger = logger

    async def OnFunctionInovationAsync(self, context: FunctionInvocationContext,
    next: Callable[[FunctionInvocationContext], Coroutine[Any, Any, None]]) -> None:
        if self.logger.level == logging.DEBUG:
            self.logger.debug(f"function invoking: {context.function.name}")
            self.logger.debug(f"arguments: {json.dumps(context.arguments)}")

        await next(context)

        if self.logger.level == logging.DEBUG:
            self.logger.debug(f"function invoked: {context.function.name}")
            self.logger.debug(f"results: \n {context.result}")
